File: examples/q8.c
/**
 * @file examples/q8.c
 *
 * @brief Prototype for QINT8 quantization.
 *
 * @note A modern transformer typically consists of 32 blocks.
 *   - A block is a single layer composed of 9 sub-layers.
 *   - Each sub-layer usually consists of one of two shapes.
 *     - @note Shapes represent row major order.
 *     - Shape 1: Vector: (n, 1) -> n (cols) x 1 (rows)
 *     - Shape 2: Matrix: (n, m) -> n (cols) x m (rows)
 * - @note F32, F16, I16, and I8 will typically contain 32 blocks.
 */

#include <assert.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define MAX_SAMPLES 10

// Quantization structure
typedef struct Quant {
    float scalar; /**< Scaling factor for quantization of input */
    uint8_t bits; /**< Quantized value */
} Quant;

typedef Quant Q8;

// 8-bit integer quantization
Q8 quantize_q8(float value) {
    Q8 q8;

    // Set the range for the integer domain [-128, 127]
    int z_domain = 255;

    // Calculate the dynamic range for the real domain [-value, value]
    float r_domain = fabsf(value); // reflect and compound input

    // if value is greater than z_max or value is less than z_min, otherwise 1
    float squeeze = 1;
    if (r_domain > z_domain) {
        squeeze = z_domain / r_domain; // ratio of the effective range
    }

    // Calculate the scaling factor and preserve the sign
    q8.scalar = (squeeze * r_domain) / z_domain;
    q8.scalar = (value >= 0) ? q8.scalar : -q8.scalar;

    // Quantize the value
    q8.bits = (uint8_t) roundf(value / q8.scalar);

    return q8;
}

// Reconstruct the real value using the scalar
float dequantize_q8(Q8 q8) {
    return (float) (q8.scalar * q8.bits); // / q8.alpha;
}

// Function to sample floating-point values in the range [-n, n-1]
void sampler(double* x, int length, int range) {
    assert(x != NULL);
    assert(length > 0);
    assert(range > 1);

    for (int i = 0; i < length; ++i) {
        double normalized = (double) rand() / (double) RAND_MAX;
        x[i] = -(range + 1) + (normalized * ((2 * range) - 1));
    }
}

// Compute absolute and relative errors
void error(double x, double x_prime, double* abs_error, double* rel_error) {
    *abs_error = fabs(x - x_prime);
    *rel_error = (fabs(x) > 1e-6) ? (*abs_error / fabs(x)) : 0.0;
}

// Stress test for validating out-of-bounds behavior
#define TEST_COUNT 3

typedef struct TestCase {
    const char* label;
    int exponent;
} TestCase;

void run_tests(int seed) {
    // initialize the rng
    srand(seed);
    printf("=== Running tests with seed: %d ===\n\n", seed);

    // Test parameters
    double input[MAX_SAMPLES];

    TestCase test_cases[TEST_COUNT] = {
        {"8-bit Signed Test",  127 },
        {"8-bit Unsigned Test", 255},
        {"16-bit Signed Test", 32535}
    };

    // Initialize error accumulators
    double total_abs_error = 0.0, total_rel_error = 0.0;
    for (int t = 0; t < TEST_COUNT; t++) {
        fprintf(
            stdout,
            "=== %s (Range: [%d, %d]) ===\n",
            test_cases[t].label,
            -(test_cases[t].exponent + 1),
            test_cases[t].exponent
        );

        // Update the sampled inputs
        sampler(input, MAX_SAMPLES, test_cases[t].exponent);
        for (int sample = 0; sample < MAX_SAMPLES; sample++) {
            // Get sample, calculate quant, and dequant
            double x = input[sample];
            Q8 q = quantize_q8(x);
            double x_prime = dequantize_q8(q);
            printf("Input: %.6f, Quant: %u, Prime: %.6f\n", x, q.bits, x_prime);

            // Calculate accumulated errors
            double abs_error, rel_error;
            error(x, x_prime, &abs_error, &rel_error);
            total_abs_error += abs_error;
            total_rel_error += rel_error;
            printf("Absolute Error: %.6f, Relative Error: %.2f%%\n\n", abs_error, rel_error * 100);
        }
    }

    printf("Average Absolute Error: %.6f\n", total_abs_error / MAX_SAMPLES);
    printf("Average Relative Error: %.2f%%\n", (total_rel_error / MAX_SAMPLES) * 100);
    printf("=== Completed %d test cases ===\n", TEST_COUNT);
}

int main(void) {
    run_tests(1337);
    return 0;
}

File: examples/q8.py
"""
Mirror the C implementation to the letter.

Only one input and one output is allowed.

Do not deviate.

Lest ye be forsaken.
"""

from dataclasses import dataclass
from random import random, seed


@dataclass
class Q8:
    scalar: float = 0.0
    alpha: float = 0.0
    bits: int = 0
    residual: float = 0.0  # Additional variable to store residual precision


def quantize_scalar_q8(value: float) -> Q8:
    q8 = Q8()

    # Define the maximum possible integer range (example: [z_min, z_max])
    z_domain: int = 255

    # Define the maximum possible real range (example: [r_min, r_max])
    r_max = abs(value)
    r_min = -r_max
    r_domain = r_max - r_min

    # Handle special case for zero input
    if r_domain == 0:
        q8.scalar = 1.0
        q8.alpha = 1.0
        q8.bits = 0
        return q8

    # Calculate alpha (squeezing ratio)
    alpha = 1.0
    if r_domain > z_domain:
        alpha = z_domain / r_domain
    q8.alpha = alpha

    # Calculate the base step size
    base_step_size = r_domain / z_domain  # Decouples the step size from the scalar

    # Quantize the value using the base step size
    q8.bits = int(round(value / base_step_size))

    # Store the residual precision
    q8.residual = value - (q8.bits * base_step_size)

    # Calculate the scalar based on the squeezed range
    q8.scalar = base_step_size * alpha

    return q8


def dequantize_scalar_q8(q8: Q8) -> float:
    # Reconstruct the real value using the quantized value, scalar, and residual
    return (q8.bits * (q8.scalar / q8.alpha)) + q8.residual


def sampler(max_elements: int, z: int) -> list[float]:
    assert max_elements > 0
    assert z >= 2

    values = []
    for i in range(max_elements):
        # Generate a random value in the range [0, 1]
        normalized: float = random()
        # Map the normalized value to the range [-n, n-1]
        values.append(-z + (normalized * (2 * z - 1)))
    return values


def main():
    # Example values to test the implementation
    seed(1337)
    values = sampler(10, 2**8 - 1)

    print("Testing Quantization and Dequantization:")
    for value in values:
        q8 = quantize_scalar_q8(value)
        reconstructed = dequantize_scalar_q8(q8)
        print(
            f"Input: {value:.6f}, Quantized: {q8.bits}, Scalar: {q8.scalar:.6f}, "
            f"Alpha: {q8.alpha:.6f}, Residual: {q8.residual:.6f}, Reconstructed: {reconstructed:.6f}"
        )
        abs_error = abs(value - reconstructed)
        rel_error = abs_error / abs(value) if abs(value) > 1e-6 else 0.0
        print(f"Absolute Error: {abs_error:.6f}, Relative Error: {rel_error:.6f}")


if __name__ == "__main__":
    main()
